modules/openxr/extensions/openxr_android_extension.cpp

```diff
@@ -68,6 +69,29 @@ void OpenXRAndroidExtension::on_before_instance_created() {
 	ERR_FAIL_COND_MSG(XR_FAILED(result), "Failed to call xrInitializeLoaderKHR");
 }
 
+void *OpenXRAndroidExtension::get_instance_create_info_next() {
+	if (create_instance_extension_available && !instance_create_info_next) {
+		JNIEnv *env = get_jni_env();
+		JavaVM *vm;
+		env->GetJavaVM(&vm);
+
+		jobject activity_object = env->NewGlobalRef(static_cast<OS_Android *>(OS::get_singleton())->get_godot_java()->get_activity());
+
+		instance_create_info_next = memalloc(sizeof(XrInstanceCreateInfoAndroidKHR));
+		*reinterpret_cast<XrInstanceCreateInfoAndroidKHR *>(instance_create_info_next) = {
```

Comment:
Note that Godot has build in memory functions/macros that should be used.

That said, look at the other extensions where we add structures, the pattern we're following is to make this a member variable of the class.

Reply:
I switched to a static instance in the cpp to match the pattern you suggested (the vulkan extension uses a static class member, which has pretty much the same properties), but can't say I love it as it ties a static instance to all instances of the class. That said, it should be reasonably safe and work, as there is at most one instance. I guess some uglyness can't be avoided, just like the leaked global reference to the activity – which I copied, so now there are two of those ;)

Note that it would not be possible to place the instance into the class without major changes, as the struct is defined within `openxr_platform.h`, whose content varies depending on the `#define`s that were set up before the *first* include in each compile unit. As long as different extensions use different parts of `openxr_platform.h`, use of  that header *from other headers* should be avoided.

I'd propose to also change the vulkan extension to match whatever pattern we decide on for the android extension, as the vulkan extension includes the header in the header already and thus might cause issues down the road. Alternatively we could also move all relevant `#define`s out of the extensions to make including `openxr_platform.h` safe from headers, though that likely comes with its own problems.

Reply:
The extensions are instanced as singletons so the variable itself can be a member variable of the class, it doesn't need to be static. Same for your JAVA instance variable. 

The extension instances are also cleaned up when the OpenXR singleton is destroyed though I suspect there might be some timing issues there so that should allow us to do any cleanup needed.

Including the needed platform headers in order to properly include OpenXR has been a pain in my behind because it screws up other parts of the system. Especially on windows once windows.h is included it causes a lot of pain.

But it seems we can't internalize this enough, I wish there was a solid way to ensure only the `.cpp` files of the extensions were required to include the needed parts of OpenXR. 

Defo welcome for suggestions but lets not let your PR hang on that improvement.

Reply:
> it doesn't need to be static. 

My comment on a static member was regarding the vulkan extension I looked at for reference, that one uses a static member for its next chain entry.

> Same for your JAVA instance variable.

I can add cleanup for the `jobject`, but would prefer to treat both instances the same (so the cleanup should also apply to the instance passed via `xrInitializeLoaderKHR` in `OpenXRAndroidExtension::on_before_instance_created`). My assumption was, however, that the missing cleanup was intentional, as the standard does not clearly state whether OpenXR assumes ownership or creates its own reference.

> Defo welcome for suggestions

Imho a void pointer in the class and `memalloc` `memfree` are a bit nicer, but I can live without that. But I think you did not want me to use those two, and have other godot memory macros to use? Either way, I'm also fine merging as-is.

> I wish there was a solid way to ensure only the .cpp files of the extensions were required to include the needed parts of OpenXR

We could go the pimpl or pure abstract route. The latter is probably nicer. But both are not in the scope of this PR imho.